### DISCUSSION TITLE
ConfidentialClientApplication long latency

### DISCUSSION THREAD

**Discussion by ritabiniash** [AzureAD/microsoft-authentication-library-for-python#441] [Q&A]
Hey, we are using `ConfidentialClientApplication` via MSAL python package. In production we see that it takes around 2 seconds (P90) to create the class. We want to avoid caching since the class is heavy. Is there a workaround or usage hack to cut the latency  time ?

**Comment by rayluo** [comment]
* Why did you say "*in production* ... it takes around 2 seconds"? Was it faster when running in your local develop machine? If this happens to be the case, you need to look into the differences in your prod and local environment setup. You could [enable debug logs](https://github.com/AzureAD/microsoft-authentication-library-for-python/blob/1.16.0/sample/confidential_client_secret_sample.py#L38-L39) to help profiling MSAL in your app.

* In general, MSAL is not currently designed to lazy load most of its prerequisites, so the initialization time is not blazingly fast. By designing MSAL as an Object-Oriented API model (rather than a function-based API), the implication is that the caller would ideally keep a long-lived `ConfidentialClientApplication` (or `PublicClientApplication`, for that matter), and stick with such a calling pattern:

  ```
  app = ConfidentialClientApplication(...)

  # Per each request
  app.acquire_token_silent(...)
  app.acquire_token_by_xyz(...)
  ... # etc.
  ```
  If you can do so, then the relatively longer initialization time should not be a real issue.

* There are situations that your entire application would be short-lived, such as a command-line script. MSAL Python recently introduced a customizable http cache mechanism, so that a CLI app can still reuse some initialization effort performed by its previous run. You can search keyword "http_cache (dict)" in our [online API Reference doc](https://msal-python.readthedocs.io/en/latest/) for details.
